File: 3회차-DP/3회차 필수문제 코드/9177.py
import sys

tc = int(sys.stdin.readline())

a = b = c = ""
dp = []

def makeString(aPos, bPos, cPos):
    if dp[aPos][bPos] != -1:
        return dp[aPos][bPos]

    if aPos == len(a):
        return b[bPos:] == c[cPos:]
    elif bPos == len(b):
        return a[aPos:] == c[cPos:]

    dp[aPos][bPos] = 0 # -1 -> 0
    

    if a[aPos] == c[cPos]:
        dp[aPos][bPos] = dp[aPos][bPos] or makeString(aPos+1, bPos, cPos+1)
    if b[bPos] == c[cPos]:
        dp[aPos][bPos] = dp[aPos][bPos] or makeString(aPos, bPos+1, cPos+1)

    return dp[aPos][bPos]



for t in range(1, tc+1):
    a, b, c = sys.stdin.readline().split()
    dp = [[-1 for _ in range(300)] for __ in range(300)]
    print("Data set " + str(t) + ": ", end="")

    if makeString(0, 0, 0) == 1:
        print("yes")
    else:
        print("no")

File: 3회차-DP/3회차 필수문제 코드/9177.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;


bool dp[300][300];
bool visited[300][300];
string a, b, c;
int alen, blen, clen;

bool makeString(int apos, int bpos, int cpos) {
    if (visited[apos][bpos]) 
        return dp[apos][bpos];

    if (apos == alen) 
        return b.substr(bpos) == c.substr(cpos);
    else if (bpos == blen) 
        return a.substr(apos) == c.substr(cpos);

    bool &ret = dp[apos][bpos];

    if (!ret && a[apos] == c[cpos]) 
        ret = ret || makeString(apos+1, bpos, cpos+1);

    if (!ret && b[bpos] == c[cpos]) 
        ret = ret || makeString(apos, bpos+1, cpos+1);

    visited[apos][bpos] = true;
    return ret;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t;
    cin >> t;

    for (int i = 1; i <= t; ++i) {
        memset(dp, false, sizeof(dp));
        memset(visited, false, sizeof(visited));

        cin >> a >> b >> c;
        alen = a.length(), blen = b.length(), clen = c.length();

        cout << "Data set " << i << ": ";
        if (makeString(0, 0, 0))
            cout << "yes" << '\n';
        else 
            cout << "no" << '\n';
    }
}